### ISSUE TITLE
layer normalization fp16 big transformer

### ISSUE THREAD

**Issue by vsuthichai** [pytorch/fairseq#335]
I've been going through the "Scaling Neural Machine Translation" paper, and according to the paper the entire forward pass happens in fp16, not including the loss calculation.  However it does include the layer norm calculation.  I'm wondering, does this not cause any reduction overflows?  My understanding of fp16 training is that ops that do large reductions such as layer norm, batch norm, softmax, etc, should be done in fp32 for safety to avoid these overflows.  Is there something else going on here to handle this when training the big transformer?

**Comment by myleott** [comment]
Good catch! We do this already for the softmaxes, but not layernorm. I’ll test this...

**Comment by vsuthichai** [comment]
@myleott I wasn't sure if the reduction was using casting somewhere behind the scenes low level.

**Comment by myleott** [comment]
It doesn't seem to make a difference for WMT En-De training with the big transformer, but is ~5% slower.

The current fairseq behavior with --fp16 is to just modify weights, inputs and optimizer, and let each model figure out for itself what individual ops to do in FP32. I'm inclined to leave these, to maximize flexibility of the models.

Separately, we'll explore adding an --apex-fp16 option that uses Nvidia's apex library as a safer alternative to --fp16, which takes care of replacing ops with safer (but slower) alternatives.
